cryptoki/src/types/mod.rs

```diff
@@ -9,606 +9,881 @@ pub mod object;
             self.flags &= !flag;
         }
     }
 
-    /// Get the TOKEN_PRESENT flag
-    pub fn token_present(&self) -> bool {
-        self.get_flag(CKF_TOKEN_PRESENT)
+    fn stringify_flag(flag: CK_FLAGS) -> &'static str {
+        match flag {
+            CKF_LIBRARY_CANT_CREATE_OS_THREADS => {
+                std::stringify!(CKF_LIBRARY_CANT_CREATE_OS_THREADS)
+            }
+            CKF_OS_LOCKING_OK => std::stringify!(CKF_OS_LOCKING_OK),
+            _ => "Unknown CK_C_INITIALIZE_ARGS flag",
+        }
     }
+}
 
-    /// Set the TOKEN_PRESENT flag
-    pub fn set_token_present(&mut self, b: bool) -> &mut Self {
-        self.set_flag(CKF_TOKEN_PRESENT, b);
-        self
+impl InitializeFlags {
+    /// Creates a new instance of `InitializeFlags` with no flags set
+    pub fn new() -> Self {
+        InitializeFlags::default()
     }
 
-    /// Get the REMOVABLE_DEVICE flag
-    pub fn removable_device(&self) -> bool {
-        self.get_flag(CKF_REMOVABLE_DEVICE)
+    /// Gets value of [`CKF_LIBRARY_CANT_CREATE_OS_THREADS`]
+    pub fn library_cant_create_os_threads(&self) -> bool {
+        self.get_flag(CKF_LIBRARY_CANT_CREATE_OS_THREADS)
     }
 
-    /// Set the REMOVABLE_DEVICE flag
-    pub fn set_removable_device(&mut self, b: bool) -> &mut Self {
-        self.set_flag(CKF_REMOVABLE_DEVICE, b);
+    /// Sets value of [`CKF_LIBRARY_CANT_CREATE_OS_THREADS`]
+    pub fn set_library_cant_create_os_threads(&mut self, b: bool) -> &mut Self {
+        self.set_flag(CKF_LIBRARY_CANT_CREATE_OS_THREADS, b);
         self
     }
 
-    /// Get the HW_SLOT flag
-    pub fn hw_slot(&self) -> bool {
-        self.get_flag(CKF_HW_SLOT)
+    /// Gets value of [`CKF_OS_LOCKING_OK`]
+    pub fn os_locking_ok(&self) -> bool {
+        self.get_flag(CKF_OS_LOCKING_OK)
     }
 
-    /// Set the HW_SLOT flag
-    pub fn set_hw_slot(&mut self, b: bool) -> &mut Self {
-        self.set_flag(CKF_HW_SLOT, b);
+    /// Sets value of [`CKF_OS_LOCKING_OK`]
+    pub fn set_os_locking_ok(&mut self, b: bool) -> &mut Self {
+        self.set_flag(CKF_OS_LOCKING_OK, b);
         self
     }
+}
 
-    /// Get the ARRAY_ATTRIBUTE flag
-    pub fn array_attribute(&self) -> bool {
-        self.get_flag(CKF_ARRAY_ATTRIBUTE)
+impl std::fmt::Display for InitializeFlags {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let flags = vec![CKF_LIBRARY_CANT_CREATE_OS_THREADS, CKF_OS_LOCKING_OK];
+        self.stringify_fmt(f, flags)
     }
+}
 
-    /// Set the ARRAY_ATTRIBUTE flag
-    pub fn set_array_attribute(&mut self, b: bool) -> &mut Self {
-        self.set_flag(CKF_ARRAY_ATTRIBUTE, b);
-        self
+impl From<InitializeFlags> for CK_FLAGS {
+    fn from(flags: InitializeFlags) -> Self {
+        flags.flags
     }
+}
 
-    /// Get the RNG flag
-    pub fn rng(&self) -> bool {
-        self.get_flag(CKF_RNG)
+impl From<CK_FLAGS> for InitializeFlags {
+    fn from(flags: CK_FLAGS) -> Self {
+        Self { flags }
     }
+}
```

Comment:
Could this be a `TryFrom` that checks that only the flags that are used by this type are set?

Reply:
Sure, I can do that.  It's _technically_ not disallowed (sorry double-negative) to pass in values outside of the range of the spec (as in, the spec doesn't directly say that if a value that isn't `CKF_LIBRARY_CANT_CREATE_OS_THREADS` or `CKF_OS_LOCKING_OK` to return `CKR_BAD_ARGUMENTS`), but it's definitely possible some implementer may error out.

There are some exception, e.g. the `flags` field of `CK_INFO` must be 0, or the argument to `C_Finalize` must be a `NULL_PTR`.

I can put in a `TryFrom`.

Reply:
I'm not opposed to doing this for the `InitializeFlags`, as it is up to the user to set them, but I don't think we should do this for other flag values that a cryptoki token may return.  The reason is that there may exist tokens that provide vendor specific flags (which is outside the PKCS#11 spec, vendors should only be defining attributes, mechanisms, key types, return types, etc., but vendors will do what vendors do) and then performing a `TryFrom` on a return may cause the function to error out, even if the P11 call succeeds.

Thoughts?

Reply:
Ah, yes, that sounds sensible. Actually, are there "bit-zones" within the integer that are vendor-controlled (as opposed to strictly controlled by the spec), and which we could ignore in the `TryFrom`?

Reply:
For thinks like `CKR_` and `CKA_` and `CKK_` and `CKM_`  and `CKO_` types, yes anything at `0x8000_0000` and above are 'vendor defined'.  Technically that doesn't exist for flags but vendors "should" be following that convention.  Maybe we have an 'VendorDefined' field in the different enums that contains values that are "unexpected".

Reply:
Makes sense. So with the `TryFrom` updated for `InitializeFlags` we're set. I'm guessing these extra, vendor-specific flags can be accessed by just dereferencing the integer value within?

Reply:
Yes.  I'm thinking of at some point added a new variant in the above mentioned `enums` to provide Vendor specific definitions